Fixes variable expansion false positive

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

<!--
Add one of the following kinds:
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

/sig testing
/sig windows

**What this PR does / why we need it**:

Some of the tests are negative test cases which are supposed to ensure that those invalid usecases are handled properly.

However, some of the tests are false positives, they can pass due to various reasons. One such example is: ``should fail substituting values in a volume subpath with absolute path``.
This test can pass if:

- the Pod cannot start due to various reasons (e.g.: the container image cannot be pulled or does
  not exist).
- the Pod ran to completion, even though the container was not supposed to start in the first place.

``filepath.IsAbs`` does not consider ``/`` or ``\`` prefixed paths on Windows as absolute, even though in the end they are treated as such. Because of this, absolute paths can be mounted as volume subpaths in Windows.

The test ``should fail substituting values in a volume subpath with absolute path`` previously passed on Windows because of a false positive in the test: the test did not specifically check that the container failed to be created, instead it checked if it was not Running (in most cases, the container was running to Completion instead, which is indeed not Running, but not correct).

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #94006

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

claudiubelu committed:
tests: Fixes variable expansion false positive test

Some of the tests are negative test cases which are supposed to ensure that those
invalid usecases are handled properly.

However, some of the tests are false positives, they can pass due to various reasons.
One such example is: "should fail substituting values in a volume subpath with absolute path".
This test can pass if:

- the Pod cannot start due to various reasons (e.g.: the container image cannot be pulled or does
  not exist).
- the Pod ran to completion, even though the container was not supposed to start in the first place.

ZhiFeng1993 commented on test/e2e/framework/pod/resource.go:
```diff
@@ -298,6 +298,26 @@ func podsRunning(c clientset.Interface, pods *v1.PodList) []error {
 	return e
 }
 
+func podContainerFailed(c clientset.Interface, namespace, podName string, containerIndex int, reason string) wait.ConditionFunc {
```
What is this containerIndex user for?

claudiubelu replied:
I was considering the possibility that a test could spawn a pod with multiple containers, so this would basically specify which one to check for. Niche usecase though.

oomichi commented on test/e2e/framework/pod/resource.go:
```diff
@@ -298,6 +298,26 @@ func podsRunning(c clientset.Interface, pods *v1.PodList) []error {
+		}
+		switch pod.Status.Phase {
+		case v1.PodPending:
+			containerStatus := pod.Status.ContainerStatuses[containerIndex]
```
This part makes an error like
```
test/e2e/framework/framework.go:597
Test Panicked
staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55

Panic: runtime error: index out of range [0] with length 0

Full stack:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x486a820, 0xc003066c00)
	GOROOT/src/runtime/panic.go:969 +0x175
k8s.io/kubernetes/test/e2e/framework/pod.podContainerFailed.func1(0xc00299b360, 0x7f6fc02cb7f0, 0xc001c75440)
	test/e2e/framework/pod/resource.go:309 +0x3c6
```

oomichi replied:
`pod.Status.ContainerStatueses` can be length 0.
So it is better to add the following check before current line 309
```
if len(status.ContainerStatuses) == 0 {
    return false, nil
}
```

claudiubelu replied:
good point, I've included this.

oomichi reviewed (commented):
/cc @oomichi

oomichi commented:
Thanks for updating.
The test part looks good for me.

/lgtm
/approve

claudiubelu commented:
/assign @derekwaynecarr

marosset commented:
/sig windows
/milestone v1.20

marosset commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -185,7 +185,10 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			}
 
-			if filepath.IsAbs(subPath) {
+			// On Windows, filepath.IsAbs does not consider \ or / prefixed paths as absolute, even though
```
I think this check is also performed at spec validation which should currently only ever happen on code compiled for linux.

https://github.com/kubernetes/kubernetes/blob/544b74c2cb94fce6d89019b85c7c891050df2af7/pkg/apis/core/validation/validation.go#L2427-L2430

calls

https://github.com/kubernetes/kubernetes/blob/544b74c2cb94fce6d89019b85c7c891050df2af7/pkg/apis/core/validation/validation.go#L1158-L1170

claudiubelu committed:
tests: Use Windows absolute path on Windows pods

The test "should fail substituting values in a volume subpath with absolute path" creates a pod
with a variable expansion path which is set as an absolute path. However "/tmp" is not an absolute
on Windows, it has to be prefixed with the drive letter (C:\tmp). But C:\tmp does not typically
exist on Windows nodes, so we use C:\Users instead.

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -185,7 +185,10 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
-			if filepath.IsAbs(subPath) {
+			// On Windows, filepath.IsAbs does not consider \ or / prefixed paths as absolute, even though
+			// they are treated as such.
+			isWindowsAbs := runtime.GOOS == "windows" && (strings.HasPrefix(subPath, `/`) || strings.HasPrefix(subPath, `\`))
```
should this logic has a separate method for validation? In a PR description I see something about other place having (or needing to have) the same logic. Would be great to cross-reference those or use the same method in both cases.

> even though in the end they are treated as such. Because of this, absolute paths can be mounted as volume subpaths in Windows.

marosset commented:
/lgtm

claudiubelu commented:
A milestone should be added to this.

marosset commented:
/milestone v1.20
/kind failing-test

justaugustus commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

BenTheElder commented:
/retest

justaugustus commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel
ref: https://github.com/kubernetes/kubernetes/issues/94263

BenTheElder commented:
/retest
